pkg/core/blockchain_test.go

```diff
@@ -2003,3 +2005,258 @@ func setSigner(tx *transaction.Transaction, h util.Uint160) {
 		Scopes:  transaction.Global,
 	}}
 }
+
+func TestBlockchain_StartFromExistingDB(t *testing.T) {
+	newPS := newLevelDBForTestingWithPath
+	ps, path := newPS(t, "")
+	var cfg *config.Config
+	customConfig := func(c *config.Config) {
+		c.ProtocolConfiguration.StateRootInHeader = true // Need for P2PStateExchangeExtensions check.
+		cfg = c
+	}
+	bc := initTestChain(t, ps, customConfig)
+	go bc.Run()
+	initBasicChain(t, bc)
+	require.True(t, bc.BlockHeight() > 5) // Ensure that basic chain is correctly initialised.
+
+	// Information for further tests.
+	h := bc.BlockHeight()
+	cryptoLibHash, err := bc.GetNativeContractScriptHash(nativenames.CryptoLib)
+	require.NoError(t, err)
+	cryptoLibState := bc.GetContractState(cryptoLibHash)
+	require.NotNil(t, cryptoLibState)
+	var (
+		managementID             = -1
+		managementContractPrefix = 8
+	)
+
+	bc.Close() // Ensure persist is done and persistent store is properly closed.
+
+	t.Run("mismatch storage version", func(t *testing.T) {
+		ps, _ = newPS(t, path)
+		t.Cleanup(func() { require.NoError(t, ps.Close()) })
+		cache := storage.NewMemCachedStore(ps) // Extra wrapper to avoid good DB corruption.
+		d := dao.NewSimple(cache, cfg.ProtocolConfiguration.StateRootInHeader, cfg.ProtocolConfiguration.P2PStateExchangeExtensions)
+		d.PutVersion(dao.Version{
+			Value: "0.0.0",
+		})
+		_, err := d.Persist() // Persist to `cache` wrapper.
+		require.NoError(t, err)
+		_, err = initTestChainNoCheck(t, cache, func(c *config.Config) {
+			customConfig(c)
+		})
+		require.Error(t, err)
+		require.True(t, strings.Contains(err.Error(), "storage version mismatch"))
+	})
+	t.Run("mismatch StateRootInHeader", func(t *testing.T) {
+		ps, _ = newPS(t, path)
+		t.Cleanup(func() { require.NoError(t, ps.Close()) })
+		_, err := initTestChainNoCheck(t, ps, func(c *config.Config) {
+			customConfig(c)
+			c.ProtocolConfiguration.StateRootInHeader = false
+		})
+		require.Error(t, err)
+		require.True(t, strings.Contains(err.Error(), "StateRootInHeader setting mismatch"))
+	})
+	t.Run("mismatch P2PSigExtensions", func(t *testing.T) {
+		ps, _ = newPS(t, path)
+		t.Cleanup(func() { require.NoError(t, ps.Close()) })
+		_, err := initTestChainNoCheck(t, ps, func(c *config.Config) {
+			customConfig(c)
+			c.ProtocolConfiguration.P2PSigExtensions = false
+		})
+		require.Error(t, err)
+		require.True(t, strings.Contains(err.Error(), "P2PSigExtensions setting mismatch"))
+	})
+	t.Run("mismatch P2PStateExchangeExtensions", func(t *testing.T) {
+		ps, _ = newPS(t, path)
+		t.Cleanup(func() { require.NoError(t, ps.Close()) })
+		_, err := initTestChainNoCheck(t, ps, func(c *config.Config) {
+			customConfig(c)
+			c.ProtocolConfiguration.StateRootInHeader = true
```

Comment:
It is done intentionally in case if one day we'll extend custom configuration with several settings. Then only one setting is being overriden in this test, while others should stay the same.